gnd/src/androidTest/java/com/google/android/gnd/BaseMainActivityTest.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd;
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
+import androidx.test.espresso.IdlingRegistry;
+import androidx.test.ext.junit.rules.ActivityScenarioRule;
+import dagger.hilt.android.testing.HiltAndroidRule;
+import javax.annotation.OverridingMethodsMustInvokeSuper;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+
+/**
+ * Configures all necessary rules and {@link MainActivity} needed for running instrumentation tests.
+ */
+public abstract class BaseMainActivityTest {
```

Comment:
Would a new test rule that includes all of these logic be better to favor composition over inheritance?

Reply:
What do you think about this suggestion?

Just want to make sure you considered this option. The code base seems to already have a bunch of abstract classes for this kind of code re-use so I guess that could be fine to have it here, too. I personally feel too much usage of inheritance could make the code a bit difficult to read because of the hidden inherited code. Using composition looks more straightforward to me from the point of a reader, so I often use composition whenever I can.

Reply:
I haven't written a lot of integration tests. So I'm not sure which is preferred. We can revert the changes for integration tests and handle them separately if you prefer. Wdyt?

Reply:
That's probably fine. Just want to check if we can agree on some best practices moving forward.

Reply:
+1 to preferring [Composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) except where inheritance provides concrete value. I think in this case since the class is inheriting annotations it might make sense, but I don't know enough about how Hilt rules work. In either case, this lgtm for now, we can circle back later if necessary.

Reply:
I tried converting it into a custom test rule. It had 2 issues:
 1. Hilt injection requires `@HiltAndroidTest` annotation. So that couldn't be put inside the test rule. [Minor issue, could be ignored by keeping this in the main test class]
 2. Tests run without waiting for the binding to complete. Needs more investigation.

Created https://github.com/google/ground-android/issues/1029 to track this separately.